src/Modules/GE Service Fabric Core/Trinity.GraphEngine.ServiceFabric.Core/GraphEngineStatefulServiceCore.cs

```diff
@@ -0,0 +1,116 @@
+using System.Runtime.Remoting.Contexts;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Trinity.Diagnostics;
+using Trinity.Network;
+using Trinity.Utilities;
+
+namespace Trinity.GraphEngine.ServiceFabric.Core
+{
+    public sealed class GraphEngineStatefulServiceCore
+    {
+        internal TrinityServer TrinityServer = null;
+        internal NodeContext NodeContext     = null;
+        internal FabricClient FabricClient   = null;
+
+        public static GraphEngineStatefulServiceCore Instance = null;
+        private static object s_lock = new object();
+
+        public List<System.Fabric.Query.Partition> Partitions { get; private set; }
+        public int PartitionCount { get; private set; }
+        public int PartitionId { get; private set; }
+        public int Port { get; private set; }
+        public int HttpPort { get; private set; }
+        public string Address { get; private set; }
+        public ReplicaRole Role { get; private set; }
+        public StatefulServiceContext Context { get; private set;}
+
+        //  Passive Singleton
+        public GraphEngineStatefulServiceCore(StatefulServiceContext context)
+        {
+            //  Initialize other fields and properties.
+            this.Context = context;
+            NodeContext  = context.NodeContext;
+            FabricClient = new FabricClient();
+            Address      = NodeContext.IPAddressOrFQDN;
+            Partitions   = FabricClient.QueryManager
+                           .GetPartitionListAsync(context.ServiceName)
+                           .GetAwaiter().GetResult()
+                           .OrderBy(p => p.PartitionInformation.Id)
+                           .ToList();
+
+            PartitionId    = Partitions.FindIndex(p => p.PartitionInformation.Id == context.PartitionId);
+            PartitionCount = Partitions.Count;
+
+            Port     = context.CodePackageActivationContext.GetEndpoint("TrinityProtocolEndpoint").Port;
+            HttpPort = context.CodePackageActivationContext.GetEndpoint("HttpEndpoint").Port;
+
+            var ags = TrinityConfig.CurrentClusterConfig.Servers;
+            ags.Clear();
+            ags.Add(new AvailabilityGroup("LOCAL", new ServerInfo("localhost", Port, null, LogLevel.Info)));
+
+            TrinityConfig.HttpPort = HttpPort;
+
+            Log.WriteLine("{0}", $"WorkingDirectory={context.CodePackageActivationContext.WorkDirectory}");
+            TrinityConfig.StorageRoot = Path.Combine(context.CodePackageActivationContext.WorkDirectory, $"P{PartitionId}{Path.GetRandomFileName()}");
+            Log.WriteLine("{0}", $"StorageRoot={TrinityConfig.StorageRoot}");
+
+            lock (s_lock)
+            {
+                if (Instance != null)
+                {
+                    Log.WriteLine(LogLevel.Fatal, "Only one GraphEngineService allowed in one process.");
+                    Thread.Sleep(5000);
+                    Environment.Exit(-1);
```

Comment:
Note: this is a quirk in current SF infra.. It sometimes tries to start multiple service cores in one process. The workaround is to simply terminate the process if this behavior is detected.